File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/ItemCategoryEnum.java
package ai.ragu.charactermanager.enumeration;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonValue;

import java.util.HashMap;
import java.util.Map;

public enum ItemCategoryEnum {
    LIGHT("light"),
    MEDIUM("medium"),
    HEAVY("heavy");

    private final String value;
    private final static Map<String, ItemCategoryEnum> CONSTANTS = new HashMap<>();

    static {
        for (ItemCategoryEnum c: values()) {
            CONSTANTS.put(c.value, c);
        }
    }

    ItemCategoryEnum(String value) {
        this.value = value;
    }

    @Override
    public String toString() {
        return this.value;
    }

    @JsonValue
    public String value() {
        return this.value;
    }

    @JsonCreator
    public static ItemCategoryEnum fromValue(String value) {
        ItemCategoryEnum constant = CONSTANTS.get(value);
        if (constant == null) {
            throw new IllegalArgumentException(value);
        } else {
            return constant;
        }
    }
}

File: charactermanager/backend/src/main/java/ai/ragu/charactermanager/enumeration/ItemCategoryEnum.kt
package ai.ragu.charactermanager.enumeration

import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonValue
import java.util.*

enum class ItemCategoryEnum(private val value: String) {
    LIGHT("light"), MEDIUM("medium"), HEAVY("heavy");

    companion object {
        private val CONSTANTS: MutableMap<String, ItemCategoryEnum> = HashMap()
        @JsonCreator
        fun fromValue(value: String?): ItemCategoryEnum {
            val constant = CONSTANTS[value]
            return constant ?: throw IllegalArgumentException(value)
        }

        init {
            for (c in values()) {
                CONSTANTS[ai.ragu.charactermanager.enumeration.c.value] = ai.ragu.charactermanager.enumeration.c
            }
        }
    }

    override fun toString(): String {
        return value
    }

    @JsonValue
    fun value(): String {
        return value
    }

}